src/le-responder/output_observer.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"encoding/hex"
+	"log"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+type certObserver interface {
+	CertsAreUpdated(certs []*credhubCert) error
+}
+
+type bucket struct {
+	Region       string `yaml:"region"`
+	Bucket       string `yaml:"bucket"`
+	Object       string `yaml:"object"`
+	AccessKey    string `yaml:"access_key"`
+	AccessSecret string `yaml:"access_secret"`
+
+	awsMutex   sync.Mutex
+	awsSession *session.Session
+
+	lastSuccessfulWritten []byte
+}
+
+func stringval(s *string) string {
+	if s == nil {
+		return "n/a"
+	}
+	return *s
+}
+
+func (b *bucket) Put(data []byte) error {
+	b.awsMutex.Lock()
+	defer b.awsMutex.Unlock()
+
+	if bytes.Equal(data, b.lastSuccessfulWritten) {
+		return nil
+	}
+
+	if b.awsSession == nil {
+		sess, err := session.NewSession(&aws.Config{
+			Region:      aws.String(b.Region),
+			Credentials: credentials.NewStaticCredentials(b.AccessKey, b.AccessSecret, ""),
+		})
+		if err != nil {
+			return err
+		}
+		b.awsSession = sess
+	}
+
+	result, err := s3manager.NewUploader(b.awsSession).Upload(&s3manager.UploadInput{
+		Bucket: aws.String(b.Bucket),
+		Key:    aws.String(b.Object),
+		Body:   bytes.NewReader(data),
+	})
+	if err != nil {
+		return err
+	}
+
+	b.lastSuccessfulWritten = data
+
+	log.Printf("Cert tarball successfully uploaded to: %s (version %s)\n", result.Location, stringval(result.VersionID))
```

Comment:
I'm confused about why this is creating a tarball.  Isn't this just managing certs?  Brendan's pipeline will be managing the release.

Reply:
Yes, Brendan's pipeline will be managing the release, but writing a single tarball out from here as input to that pipeline makes it easier to reason about when a complete set of certs is ready or not, and I would guess might make it easy to use some built-in S3 object notification stuff to get that trigger.

Reply:
There are some advantages to making the directory layout of the bucket match the release and the config on the server.  For one, it makes things a lot easier to debug (e.g., we can check for problems just by diffing).  For another, it helps us make totally decoupled components.  The release pipeline doesn't have to know anything about how HAProxy is configured or how certs are generated, and the cert generator doesn't have to know anything about the deployment process, and the HAProxy BOSH release doesn't even have to know anything about how our configuration works (it just loads it up).  This decoupling makes it easier to use third-party tooling, or add new features to one part of the system.

If you want to make a trigger for the pipeline to reload certs, you could just write a separate object ("release.txt" or something).  It could contain a list of updated objects if you want, but I'm not sure if we'd use it.